daft/filesystem.py

```diff
@@ -15,10 +16,33 @@
 
 import fsspec
 import pyarrow as pa
+from fsspec.implementations.http import HTTPFileSystem
 from loguru import logger
+from pyarrow.fs import FileSystem
 
 from daft.datasources import ParquetSourceInfo, SourceInfo
 
+_LOCAL_SCHEME = "local"
+_CACHED_FSES: dict[str, FileSystem] = {}
+
+
+def _get_fs_from_cache(protocol: str) -> FileSystem | None:
```

Comment:
Can't we leverage `functools` LRU cache here instead of rolling our own singleton?

Reply:
Not in a straightforward way:
- We're delegating to pyarrow for translating a URI to a filesystem + a "resolved" URI since that saves us from reimplementing a ton of logic (e.g. mapping S3 URI query parameters to `S3FileSystem` configurations).
- Nearly all URIs will be unique, so we can't rely on those being reasonable cache keys.

Besides, if you look at the CPython implementation, `functools.lru_cache(maxsize=None)` is just a light wrapper around a Python dictionary, so we're not getting much benefit from it.

Reply:
Oh... yeah I was thinking:

```
@functools.lru_cache(maxsize=None)
def _get_fs_from_cache(protocol: str) -> ...
    ...
```

So we can cache on the protocol, not the unique URI

Reply:
As mentioned in [another comment](https://github.com/Eventual-Inc/Daft/pull/942#issuecomment-1555663518), if we're wanting to leverage pyarrow's URI --> filesystem logic, we need to provide the full URI, not just the protocol. We could start hosting this logic ourselves, but then we'd need to also do the work of translating URI query parameters and the like to filesystem configurations, which pyarrow already does for us. We would also need to deal with the different support for storage backends across different pyarrow versions, which may be messy.